pkg/provider/azure_controller_common.go

```diff
@@ -267,7 +279,38 @@ func (c *controllerCommon) AttachDisk(ctx context.Context, async bool, diskName,
 			klog.Warningf("azureDisk - switch to batch operation due to rate limited, QPS: %f", c.diskOpRateLimiter.QPS())
 		}
 	}
-	return lun, vmset.WaitForUpdateResult(ctx, future, nodeName, "attach_disk")
+
+	backoffManager := wait.Backoff{
+		Duration: updateVMUpdateRetryDuration,
+		Factor:   updateVMUpdateRetryFactor,
+		Steps:    updateVMUpdateRetrySteps,
+	}
+
+	err = vmset.WaitForUpdateResult(ctx, future, nodeName, "attach_disk")
+	for c.vmUpdateRequired(future, err) {
```

Comment:
there is already `UpdateVM` func, I think you could use that function directly, and for disk attach/detach, it does not need backoff, if it failed once, then return error directly, CSI driver has backoff mechanism, it would retry next.

https://github.com/kubernetes-sigs/cloud-provider-azure/blob/063aa0f8e2a12f11b354a20a188cae2af45000b1/pkg/provider/azure_controller_common.go#L375-L385

Reply:
The purpose of this PR is to address a customer reported issue where a retry triggered by csi driver fails to handle `OperationPreempted` error due to `VM Provisioning` state not being in failed state. After discussing the scenario with the VMSS team, we found it more appropriate to retry on certain errors (in this case, specifically `OperationPreempted` error) in cloud-provider-azure layer. And we needed an async version of `UpdateVM` function as the code change has dependency on `future` value.

Reply:
The current change only retries if `OperationPreempted` error occurs. Or are you suggesting that we get rid of retry in whole? 

The reason we cannot rely on CSI driver retry and backoff is that from what we have gathered from customer reported incidence discussion, we have no guarantee that VM provisioning state would be in `failed` state on the second `ControllerPublishVolume` call, in which case I think the driver returns success as the disk lun had already been assigned and commited for the disk. That is why we are trying to retry in `cloud-provider-azure` level, for which case we would need a back off implemented as the next VM update can fail with the same error.

Do you have any specific concerns or preferred way to address this issue?